File: alignment/needleman_wunsch.py
from ._alignment import Alignment


class NeedlemanWunsch(Alignment):
    def _align(self):
        seq1 = " " + self.seq1
        seq2 = " " + self.seq2
        delta = self.delta
        matrix = [[- delta * i for i in range(len(seq2))]]
        
        for i in range(1, len(seq1)):
            row = [- delta * i]
            
            for j in range(1, len(seq2)):
                Sa = matrix[i-1][j-1] + self.subs_mat[seq1[i]][seq2[j]]
                Sb = matrix[i-1][j] - delta
                Sc = row[j-1] - delta
                
                Sij = max(Sa, Sb, Sc)
                row.append(Sij)
            matrix.append(row)
        self.align_matrix = matrix
    
    def _trace_back(self):
        matrix = self.align_matrix
        seq1, seq2 = " " + self.seq1, " " + self.seq2
        i = len(matrix) - 1
        j = len(matrix[0]) - 1

        alseq1 = []
        alseq2 = []
        matches = []

        while i > 0 and j > 0:
            diag = matrix[i-1][j-1]
            up = matrix[i-1][j-1]
            left = matrix[i][j-1]
            best = max(diag, up, left)

            if best == diag:
                alseq1.append(seq1[i])
                alseq2.append(seq2[j])
                matches.append("|" if seq1[i] == seq2[j] else " ")
                i -= 1
                j -= 1
            elif best == up:
                alseq1.append(seq1[i])
                alseq2.append("-")
                matches.append(" ")
                i -= 1
            else:
                alseq1.append("-")
                alseq2.append(seq2[j])
                matches.append(" ")
                j -= 1
        self.alseq1 = "".join(alseq1[::-1])
        self.alseq2 = "".join(alseq2[::-1])
        self.almatches = "".join(matches[::-1])

File: cpp/alignment/NeedlemanWunsch.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include "NeedlemanWunsch.h"

void NeedlemanWunsch::align() {
    auto & matrix = this->matrix();
    auto & subMat = getSubMat();
    std::string const* seq1 = getSeq1();
    std::string const* seq2 = getSeq2();
    double delta = getDelta();

    for(int i = 0; i < seq1->length(); ++i) {
        std::vector<double> row;
        for(int j = 0; j < seq2->length(); ++j) {
            if(i == 0) {
                double Sij = - delta * j;
                row.push_back(Sij);
            } else if (j == 0) {
                double Sij = - delta * i;
                row.push_back(Sij);
            } else {
                double Sa = matrix[i-1][j-1] + 
                    subMat.at(seq1->substr(i, 1)).at(seq2->substr(j,1));
                double Sb = matrix[i-1][j] - delta;
                double Sc = row[j-1] - delta;

                double Sij = std::max(Sa, Sb);
                Sij = std::max(Sij, Sc);
                row.push_back(Sij);
            }
        }
        matrix.push_back(row);
    }
}

void NeedlemanWunsch::trace_back() {
    auto & matrix = this->matrix();
    std::string const* seq1 = getSeq1();
    std::string const* seq2 = getSeq2();

    int i = matrix.size() - 1;
    int j = matrix[0].size() - 1;

    std::string alseq1, alseq2, matches;

    while(i > 0 && j > 0) {
        double diag = matrix[i - 1][j - 1];
        double up = matrix[i - 1][j - 1];
        double left = matrix[i - 1][j - 1];
        double best = std::max(diag, up);
        best = std::max(best, left);

        if (best == diag){
            alseq1.push_back((*seq1)[i]);
            alseq2.push_back((*seq2)[j]);
            matches.push_back(seq1->substr(i, 1) == seq2->substr(j, 1) ? '|' : ' ');
            --i;
            --j;
        } else if (best == up){
            alseq1.push_back((*seq1)[i]);
            alseq2.push_back('-');
            matches.push_back(' ');
            --i;
        } else {
            alseq1.push_back('-');
            alseq2.push_back((*seq2)[j]);
            matches.push_back(' ');
            --j;
        }
    }
    std::reverse(alseq1.begin(), alseq1.end());
    std::reverse(alseq2.begin(), alseq2.end());
    std::reverse(matches.begin(), matches.end());
    setAlignedSeqs(alseq1, alseq2, matches);
}